DirectProgramming/DPC++FPGA/PlatformTests/board_test/src/board_test.hpp

```diff
@@ -0,0 +1,1852 @@
+              << " MB buffer\n\n";
+  } else {
+    std::cout << "Wrote and readback " << (max_alloc_size / kMB)
+              << " MB buffer\n";
+    std::cout
+        << "Failed write/readback test with " << errors << " errors out of "
+        << compare_count << " \\ 0x" << std::hex << compare_count
+        << std::dec  // Restoring manipulator to decimal at the end of cout
+        << " comparisons\n\n";
+  }
+
+  // Free allocated host memory
+  free(hostbuf);
+  return errors;
+}
+
+//////////////////////////////////
+// **** HostSpeed function **** //
+//////////////////////////////////
+
+// Inputs:
+// queue &q - queue to submit operation
+// Returns:
+// 0 is test passes, 1 if test fails
+
+// The function does the following tasks:
+// 1. Test entire device global memory by writing and reading to it
+// 2. Write to device global memory in smaller chunks and measure transfer time
+// for each chunk
+// 3. Read back data from device global memory and measure transfer time for
+// each chunk
+// 4. Verify data read back from device matches data written to device
+// 5. Calculate write bandwidth, read bandwidth and read-write bandwidth from
+// results of above transfers
+
+// Following additional functions are used for the above tasks:
+// More details about these funtions can be found with the corresponsing
+// definitions
+// 1. size_t TestGlobalMem(queue &q)
+// 2. struct Speed WriteSpeed(queue &q, buffer<char,1> &device_buffer, char
+// *hostbuf_wr, size_t block_bytes, size_t total_bytes)
+// 3. struct Speed ReadSpeed(queue &q, buffer<char,1> &device_buffer, char
+// *hostbuf_rd, size_t block_bytes, size_t total_bytes)
+// 4. bool CheckResults
+// 4. unsigned long SyclGetQSubExecTimeNs(event e)
+// 5. unsigned long SyclGetTotalTimeNs(event first_evt, event last_evt)
+
+int ShimMetrics::HostSpeed(queue &q) {
+  // Total bytes to transfer
+  constexpr size_t kMaxBytes = 8 * kMB;  // 8 MB;
+  constexpr size_t kMaxChars = kMaxBytes / sizeof(char);
+
+  // Block size of each transfer in bytes
+  constexpr size_t kMinBytes = 32 * kKB;  // 32 KB
+  size_t block_bytes = kMinBytes;
+
+  // Call function to verify write to and read from the entire device global
+  // memory
+  if (ShimMetrics::TestGlobalMem(q) != 0) {
+    std::cout << "Error: Global memory test failed\n";
+    return 1;
+  }
+
+  // **** Device buffer **** //
+
+  // Creating device buffer to span kMaxBytes size
+  // Buffer that WriteSpeed and ReadSpeed functions write to & read from
+  buffer<char, 1> device_buffer{range<1>{kMaxChars}};
+
+  // **** Host memory allocation and initialization **** //
+
+  // hostbuf_wr is used by WriteSpeed function to get input data to device
+  // buffer hostbuf_rd is used by ReadSpeed function to store data read from
+  // device buffer
+  char *hostbuf_rd = (char *)malloc(kMaxBytes);
+  char *hostbuf_wr = (char *)malloc(kMaxBytes);
+
+  // Initializing input on host to be written to device buffer
+  srand(time(NULL));
```

Comment:
Maybe initialize with a configurable seed rather than time(NULL), so we can more easily reproduced failures.